cpu/samd21/cpu.c

```diff
@@ -10,57 +10,89 @@
-    NVIC_SetPriority(PendSV_IRQn, 0xff);
-
-    /* Initialise clock sources and generic clocks */
-    clk_init();
-}
+    PM->APBAMASK.reg = (PM_APBAMASK_PM | PM_APBAMASK_SYSCTRL |
+                        PM_APBAMASK_GCLK);
 
-/**
- * @brief Initialise clock sources and generic clocks
- */
-void clk_init(void)
-{
-    PM->APBAMASK.reg |= PM_APBAMASK_SYSCTRL;
-    PM->APBAMASK.reg |= PM_APBAMASK_GCLK;
+    /* adjust NVM wait states, see table 42.30 (p. 1070) in the datasheet */
+#if (CLOCK_CORECLOCK > 24000000)
+    PM->APBAMASK.reg |= PM_AHBMASK_NVMCTRL;
+    NVMCTRL->CTRLB.reg |= NVMCTRL_CTRLB_RWS(1);
+    PM->APBAMASK.reg &= ~PM_AHBMASK_NVMCTRL;
+#endif
 
+    /* configure internal 8MHz oscillator to run without prescaler */
     SYSCTRL->OSC8M.bit.PRESC = 0;
-    SYSCTRL->OSC8M.bit.ONDEMAND = 1;
+    SYSCTRL->OSC8M.bit.ONDEMAND = 0;
     SYSCTRL->OSC8M.bit.RUNSTDBY = 0;
     SYSCTRL->OSC8M.bit.ENABLE = 1;
+    while (!(SYSCTRL->PCLKSR.reg & SYSCTRL_PCLKSR_OSC8MRDY));
 
-    /* Software reset the module to ensure it is re-initialized correctly */
+#if CLOCK_USE_PLL
+    /* reset the GCLK module so it is in a known state */
     GCLK->CTRL.reg = GCLK_CTRL_SWRST;
-    while (GCLK->CTRL.reg & GCLK_CTRL_SWRST);
-
     while (GCLK->STATUS.reg & GCLK_STATUS_SYNCBUSY);
-    /* Select the correct generator */
-    *((uint8_t*)&GCLK->GENDIV.reg) = 0;
+
+    /* setup generic clock 1 to feed DPLL with 1MHz */
+    GCLK->GENDIV.reg = (GCLK_GENDIV_DIV(8) |
+                        GCLK_GENDIV_ID(1));
+    GCLK->GENCTRL.reg = (GCLK_GENCTRL_GENEN |
+                         GCLK_GENCTRL_SRC_OSC8M |
+                         GCLK_GENCTRL_ID(1));
+    GCLK->CLKCTRL.reg = (GCLK_CLKCTRL_GEN(1) |
+                         GCLK_CLKCTRL_ID(1) |
+                         GCLK_CLKCTRL_CLKEN);
     while (GCLK->STATUS.reg & GCLK_STATUS_SYNCBUSY);
-    GCLK->GENCTRL.reg = (GCLK_GENCTRL_GENEN | GCLK_GENCTRL_SRC_OSC8M | GCLK_CLKCTRL_GEN_GCLK0);
+
+    /* enable PLL */
+    SYSCTRL->DPLLRATIO.reg = (SYSCTRL_DPLLRATIO_LDR(CLOCK_PLL_MUL));
+    SYSCTRL->DPLLCTRLB.reg = (SYSCTRL_DPLLCTRLB_REFCLK_GCLK);
+    SYSCTRL->DPLLCTRLA.reg = (SYSCTRL_DPLLCTRLA_ENABLE);
+    while(!(SYSCTRL->DPLLSTATUS.reg &
+           (SYSCTRL_DPLLSTATUS_CLKRDY | SYSCTRL_DPLLSTATUS_LOCK)) ==
+           (SYSCTRL_DPLLSTATUS_CLKRDY | SYSCTRL_DPLLSTATUS_LOCK));
+
+    /* select the PLL as source for clock generator 0 (CPU core clock) */
+    GCLK->GENDIV.reg =  (GCLK_GENDIV_DIV(CLOCK_PLL_DIV) |
+                        GCLK_GENDIV_ID(0));
+    GCLK->GENCTRL.reg = (GCLK_GENCTRL_GENEN |
+                         GCLK_GENCTRL_SRC_FDPLL96M |
+                         GCLK_GENCTRL_ID(0));
+#else /* do not use PLL, use internal 8MHz oscillator directly */
+    GCLK->GENDIV.reg =  (GCLK_GENDIV_DIV(CLOCK_DIV) |
+                        GCLK_GENDIV_ID(0));
+    GCLK->GENCTRL.reg = (GCLK_GENCTRL_GENEN |
+                         GCLK_GENCTRL_SRC_OSC8M |
+                         GCLK_GENCTRL_ID(0));
```

Comment:
Either that, but it would mean to use GCLK1 also with the internal oscillator and so would increase the power usage (although I doubt its really measurable...).